rfd/0001-tooling-requirements.md

```diff
@@ -0,0 +1,151 @@
+```
+/
+├── .github/
+│   ├── PULL_REQUEST_TEMPLATE/
+│   ├── renovate/
+│   │   ├── labels.json5
+│   │   └── ...
+│   ├── workflows/
+│   │   ├── some-tool-cd.yaml -> ../../tools/some-tool/workflows/cd.yaml
+│   │   ├── some-tool-ci.yaml -> ../../tools/some-tool/workflows/ci.yaml
+│   │   ├── ...
+│   ├── CODEOWNERS
+│   ├── dependabot.yml
+│   ├── renovate-repo-config.js
+│   └── renovate.json5
+├── bot/
+├── libs/
+│   ├── some-library/
+│   │   ├── docs/
+│   │   ├── workflows/
+│   │   │   ├── ci.yaml
+│   │   │   └── ...
+│   │   ├── .gitignore
+│   │   ├── CHANGELOG.md
+│   │   ├── README.md
+│   │   └── renovate.json5
+│   └── ...
+├── rfd/
+│   ├── 0001-tooling-requirements.md
+│   └── ...
+├── tools/
+│   ├── some-tool/
+│   │   ├── docs/
+│   │   ├── workflows/
+│   │   │   ├── cd.yaml
+│   │   │   ├── ci.yaml
+│   │   │   └── ...
+│   │   ├── .gitignore
+│   │   ├── CHANGELOG.md
+│   │   ├── README.md
+│   │   ├── action.yaml
+│   │   └── renovate.json5
+│   └── ...
+├── LICENSE
+├── README.md
+└── SECURITY.md
+```
+
+There are few things to note here:
+* Projects should be separated into `tools` and `libs` directories as appropriate. These directories should contain all source code, CI/CD pipelines, documentation, and dependency management configuration associated with each project. The specific layout of source code within these directories is left up to the project's code owner(s).
+* Workflows will live in tool and library directories rather than `.github/workflows` at the root of the repo. This will help ensure that it is clear which workflows correspond with each tool and library. Github still requires that all workflows live in the repo root, so they will be symlinked in `.github/workflows/<project>-<workflow>.yaml` targeting `<lib|tooling>/<project>/workflows/workflow>.yaml` instead.
+* While Dependabot has historically been used for keeping this repo's tool's dependencies up to date, Renovate will now be added as well. See the [Dependency management](#dependency-management) section for details on why this choice was made. The structure of the configuration will be similar to `gravitational/cloud-terraform`, with a top-level config and individual configs for each tool. This allows Renovate to be specifically configured for each tool.
+* There will be a pull request template for new projects that includes a checklist of all the items listed in [Project requirements](#project-requirements).
+* Some additional boilerplate files will be added such as a `LICENSE` and `SECURITY.md` file. The contents of these files will be copy/pasted from `gravitational/teleport` and tweaked as appropriate.
+
+### Project requirements
+New "projects" (tools and libraries) should live in dedicated directories under `tools` and `libs`, wherever is most appropriate. Each project directory should be treated as it's own separate repo in that all resources associated with the project's lifecycle should be self-contained in the directory.
```

Comment:
I am anticipating that there will be some `tool`s that depend on `libs`. Are we expecting to have one top-level `go.mod` for both? Or will that be managed by revovate?

...or am I completely barking up the wrong tree?

Reply:
Good catch - I should have included that. I'd like each tool and library to have their own go.mod, with renovate managing them. Most of the time the dependencies should be the same, but with this structure we should be able to move dependencies forward for most projects even if one project has to be kept behind.

I don't think this will generate an abnormally large number of dependency update PRs as long as we configure Renovate properly to group dependency updates for each project into a single PR.

Reply:
Yeah - seems fair. It would definitely be good to avoid the trap of requiring one module spec for the whole lot and painting ourselves into a corner.

Reply:
Updated to show go.mod/go.sum in the example file tree.